rails_programming/rails_basics/controller_basics.md

```diff
@@ -222,3 +214,14 @@ This section contains helpful links to other content. It isn't required, so cons
 
 * [Controller & Routes Video Demo](https://vimeo.com/168501163)
 * [Rails 3 Rendering and Partials via YouTube](http://www.youtube.com/watch?v=m-tw2OCHPMI)
+
+### Knowledge Check
+This section contains questions for you to check your understanding of this lesson. If you’re having trouble answering the questions below on your own, review the material above to find the answer.
+
+* <a class="knowledge-check-link" href="https://guides.rubyonrails.org/action_controller_overview.html#what-does-a-controller-do-questionmark">What does a controller do?</a>
+* <a class="knowledge-check-link" href="#naming-matters">Why is it important to adhere to the Rails naming convention for your controllers and all of its methods?</a>
+* <a class="knowledge-check-link" href="#rendering-and-redirecting">What is the difference between `render` and `redirect_to`?</a>
+* <a class="knowledge-check-link" href="https://guides.rubyonrails.org/action_controller_overview.html#parameters">Which variable in your controller contains all the data sent in by the user?</a>
+* <a class="knowledge-check-link" href="#strong-parameters">Data from a client can never be trusted. Which helper method should you incorporate in your controller to prevent malicious data injections?</a>
+* <a class="knowledge-check-link" href="https://guides.rubyonrails.org/action_controller_overview.html#the-flash">When is a flash message shown?</a>
+* <a class="knowledge-check-link" href="https://guides.rubyonrails.org/action_controller_overview.html#flash-now">In which case could the timing of a flash message be a problem and how do we adjust it?</a>
```

Comment:
```suggestion
* <a class="knowledge-check-link" href="https://guides.rubyonrails.org/action_controller_overview.html#flash-now">How do we change when a flash is shown?</a>
```

Reply:
Specifically, I suggest this edit so that there's only one "question" in the question.

Reply:
Approved.